library/src/main/scala/scala/scalajs/js/JSConverters.scala

```diff
@@ -51,4 +54,33 @@ object JSConverters {
       arr: scala.Array[T]): JSRichGenTraversableOnce[T] =
     new JSRichGenTraversableOnce(arr)
 
+  implicit class JSRichFuture[A](val self: Future[A]) {
+    /** Converts the Future to a JavaScript [[Promise]].
+     *
+     *  Attention! The nature of the [[Promise]] class, from the ECMAScript
+     *  specification, makes this method inherently un-typeable, because it is
+     *  not type parametric.
+     *
+     *  The signature of the `toJSPromise` method is only valid
+     *  <i>provided that</i> the values of `A` do not have a `then` method.
+     */
+    def toJSPromise: Promise[A] = {
```

Comment:
So you want:

* `(f: Future[Thenable[A]]).toJSPromise` typed as `js.Promise[A]`, and
* `(f: Future[A]).toJSPromise` typed as `js.Promise[A]`?

I guess that makes sense. I'll try to come up with some implicit voodoo.